src/query/sql/src/planner/optimizer/rule/rewrite/rule_filter_nulls.rs

```diff
@@ -0,0 +1,191 @@
+use crate::plans::FunctionCall;
+use crate::plans::Join;
+use crate::plans::JoinType;
+use crate::plans::RelOp;
+use crate::plans::RelOperator;
+use crate::ScalarExpr;
+
+const NULL_THRESHOLD_RATIO: f64 = 0.02;
+
+/// A rule that filters out NULL values from join keys when they exceed a certain threshold.
+/// This optimization can improve join performance by reducing the amount of data processed.
+pub struct RuleFilterNulls {
+    id: RuleID,
+    matchers: Vec<Matcher>,
+}
+
+impl RuleFilterNulls {
+    pub fn new() -> Self {
+        Self {
+            id: RuleID::FilterNulls,
+            // Join
+            // /  \
+            //... ...
+            matchers: vec![Matcher::MatchOp {
+                op_type: RelOp::Join,
+                children: vec![Matcher::Leaf, Matcher::Leaf],
+            }],
+        }
+    }
+
+    /// Checks if null filtering should be applied to a join key based on statistics.
+    ///
+    /// # Arguments
+    /// * `key_expr` - The join key expression to check
+    /// * `stats` - Statistics for the relation containing the key
+    /// * `cardinality` - Total cardinality of the relation
+    ///
+    /// # Returns
+    /// Some(filter) if null filtering should be applied, None otherwise
+    fn should_filter_nulls(
+        key_expr: &ScalarExpr,
+        stats: &Statistics,
+        cardinality: f64,
+    ) -> Option<ScalarExpr> {
+        if !key_expr.has_one_column_ref() {
+            return None;
+        }
+
+        let column_id = *key_expr.used_columns().iter().next()?;
+        let column_stats = stats.column_stats.get(&column_id)?;
+
+        if cardinality == 0.0 {
+            return None;
+        }
+
+        if (column_stats.null_count as f64 / cardinality) >= NULL_THRESHOLD_RATIO {
+            Some(join_key_null_filter(key_expr))
+        } else {
+            None
+        }
+    }
+}
+
+impl Rule for RuleFilterNulls {
+    fn id(&self) -> RuleID {
+        self.id
+    }
+
+    fn apply(&self, s_expr: &SExpr, state: &mut TransformResult) -> Result<()> {
+        let join: Join = s_expr.plan().clone().try_into()?;
+        if !matches!(
+            join.join_type,
+            JoinType::Inner | JoinType::LeftSemi | JoinType::RightSemi
+        ) || join.is_lateral
+        {
+            state.add_result(s_expr.clone());
+            return Ok(());
+        }
+        let mut left_child = s_expr.child(0)?.clone();
+        let mut right_child = s_expr.child(1)?.clone();
+
+        let left_stat = RelExpr::with_s_expr(&left_child).derive_cardinality()?;
+        let right_stat = RelExpr::with_s_expr(&right_child).derive_cardinality()?;
+        let mut left_null_predicates = vec![];
+        let mut right_null_predicates = vec![];
+        for join_key in join.equi_conditions.iter() {
+            let left_key = &join_key.left;
+            let right_key = &join_key.right;
+            if single_plan(&left_child) {
+                if let Some(filter) = Self::should_filter_nulls(
```

Comment:
In some cases we can eliminate `is_not_null`, for example the filter is `is_not_null(a) and a > 3`, `a > 3` will filter out null values.

Reply:
Will be implemented in future PR.